File: Programmers/seongyong/땅따먹기.py
def solution(land):
    table: list[list] = [[0 for _ in range(len(land[0]))]
                         for _ in range(len(land))]
    table[0] = land[0]
    for r in range(1, len(land)):
        for c in range(len(land[r])):
            max_in_row = 0
            for i in range(len(table[r - 1])):
                if i != c: max_in_row = max(table[r - 1][i], max_in_row)

            table[r][c] = land[r][c] + max_in_row

    return max(table[-1])

File: Programmers/hyeonyoung/땅따먹기.cpp
#include <iostream>
#include <vector>

using namespace std;

int solution(vector<vector<int>> land)
{
    int dp[2][4];
    dp[0][0] = land[0][0];
    dp[0][1] = land[0][1];
    dp[0][2] = land[0][2];
    dp[0][3] = land[0][3];

    int R = land.size();
    for (int i = 1; i < R; ++i)
    {
        for (int j = 0; j < 4; ++j)
        {
            int score = 0;
            for (int jj = 0; jj < 4; ++jj)
            {
                if (j == jj)
                {
                    continue;
                }

                score = max(score, dp[(i + 1) & 1][jj]);
            }
            dp[i & 1][j] = score + land[i][j];
        }
    }

    int answer = 0;
    for (int j = 0; j < 4; ++j)
    {
        answer = max(answer, dp[(R + 1) & 1][j]);
    }
    return answer;
}
